src/ApplicationSampler.cpp

```diff
@@ -407,10 +409,18 @@ namespace geopm
                                                                   core_idx);
                 GEOPM_DEBUG_ASSERT(inactive_cpu.size() != 0,
                                    "Valid core index returned no nested CPUs");
-                result = *(inactive_cpu.begin());
+                result = *(inactive_cpu.rbegin());
+                found_inactive_core = true;
                 break;
             }
         }
+        if (!found_inactive_core) {
+            // If no inactive cores are found, pin the Controller away from the OS.
+            result = num_cpu - 1;
+        }
+#ifdef GEOPM_DEBUG
+        std::cout << "Info: <geopm> ApplicationSampler::sampler_cpu(): The Controller will run on logical CPU " << result << std::endl;
```

Comment:
We will have to be careful about changing this message if PlatformIO owns the ApplicationSampler but is not the Controller thread.